ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,467 @@
+        // check that length is either multiple of 256 or 256 is multiple of length
+        debug_assert!(x_left.len() % 256 == 0 || 256 % x_left.len() == 0);
+        // asserts passed, create struct
+        Self {
+            x_left,
+            x_right,
+            y_left,
+            y_right,
+            prss_left,
+            prss_right,
+            z_right,
+        }
+    }
+
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.x_left.len()
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.x_left.is_empty()
+    }
+}
+
+/// `SegmentEntry` is a simple wrapper to represent one entry of a Segment
+/// currently, we only support `BitSlices`
+#[derive(Clone, Debug)]
+pub struct SegmentEntry<'a>(&'a BitSlice<u8, Lsb0>);
+
+impl<'a> SegmentEntry<'a> {
+    #[must_use]
+    pub fn new(entry: &'a BitSlice<u8, Lsb0>) -> Self {
+        SegmentEntry(entry)
+    }
+
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+/// `UnverifiedValuesStore` stores a vector of `UnverifiedValues` together with an `offset`
+/// `offset` is the minimum `RecordId` for the current batch within the gate
+#[derive(Clone, Debug)]
+struct UnverifiedValuesStore {
+    offset: RecordId,
+    vec: Vec<Option<UnverifiedValues>>,
+}
+
+impl UnverifiedValuesStore {
+    /// creates a new store for given length and offset
+    fn initialize(offset: RecordId, length: usize) -> Self {
+        Self {
+            offset,
+            vec: vec![None; length],
+        }
+    }
+
+    /// `reset_offset` reallocates vector such that a smaller `RecordId` than `offset` can be stored
+    /// `new_offset` will be the new `offset` and the store and store elements with `RecordId` `new_offset` and larger
+    /// when `new_offset` is larger that the current `offset`, the function does nothing
+    fn reset_offset(&mut self, new_offset: RecordId, segment_length: usize) {
+        if new_offset < self.offset {
+            let mut new_offset = usize::from(new_offset);
+            // if segment_length is less than 256, redefine new offset such that we don't need to rearrange the existing segments
+            if 256 % segment_length == 0 {
+                // distance between new_offset and old offset is rounded up to a multiple of 256
+                new_offset = ((usize::from(self.offset) - new_offset + 255) >> 8) << 8;
+            }
+            let extension_length = ((usize::from(self.offset) - new_offset) >> 8) * segment_length;
+            // use existing vec and add enough space in front
+            let vec_previous = &mut self.vec;
+            let mut vec: Vec<Option<UnverifiedValues>> =
+                Vec::with_capacity(vec_previous.len() + extension_length);
+            vec.resize(extension_length, None);
+            vec.append(vec_previous);
+            self.offset = RecordId::from(new_offset);
```

Comment:
Can you tell me a bit more about when you expect to call this function? Why would we need to support pre-pending elements? Why not just support adding to the end?

Reply:
I don't think you need it as long as iterate from a low record_id to a high record_id. If you don't follow that order this function makes sure you don't panic. However, it won't be great for efficiency reasons.

Reply:
More background: We did not want to use gate,record_id as key to the hashmap since there are a lot of records and this would be very expensive. Instead, we arrange the records in a vector where record_id is the index. The vector will have an offset - which is a record_id - that will be stored at index 0. If the offset is set incorrectly and you try to store a record smaller than offset, you would need to redefine the offset and reallocate the vector.

Reply:
For the test, I want to make sure that it runs without panics since efficiency is not the main concern. There are several aspects of the DZKPs that we could optimize (Optimizations that I have ignored so far) if the DZKPs turns out to be a bottleneck.